src/load_cell_endstop.c

```diff
@@ -0,0 +1,297 @@
+{
+    lce->flags &= ~mask;
+}
+
+void
+try_trigger(struct load_cell_endstop *lce, uint32_t ticks)
+{
+    // set live trigger flag for QUERY_ENDSTOPS
+    set_flag(FLAG_IS_TRIGGERED, lce);
+
+    uint8_t is_homing = is_flag_set(FLAG_IS_HOMING, lce);
+    uint8_t is_homing_triggered = is_flag_set(FLAG_IS_HOMING_TRIGGER, lce);
+
+    if (is_homing && !is_homing_triggered) {
+        // the first triggering sample when homing sets the trigger time
+        lce->trigger_ticks = ticks;
+        // this flag latches until a reset, disabling further triggering
+        set_flag(FLAG_IS_HOMING_TRIGGER, lce);
+        trsync_do_trigger(lce->ts, lce->trigger_reason);
+    }
+}
+
+void
+try_trigger_error(struct load_cell_endstop *lce)
+{
+    uint8_t is_homing = is_flag_set(FLAG_IS_HOMING, lce);
+    if (is_homing) {
+        trsync_do_trigger(lce->ts, lce->error_reason);
+    }
+}
+
+// Used by Sensors to report new raw ADC sample
+void
+load_cell_endstop_report_sample(struct load_cell_endstop *lce
+                                , const int32_t sample, const uint32_t ticks)
+{
+    // save new sample
+    lce->last_sample = sample;
+    lce->last_sample_ticks = ticks;
+    lce->watchdog_count = 0;
+    // check for safety limit violations first
+    const uint8_t is_safety_trigger = sample <= lce->safety_counts_min
+                                        || sample >= lce->safety_counts_max;
+    // too much force, this is an error while homing
+    if (is_safety_trigger) {
+        try_trigger_error(lce);
+    }
+
+    // only trigger when trigger_grams is greater than 0
+    if (lce->trigger_grams == 0) {
+        // clear "live" trigger view for QUERY_ENDSTOPS
+        clear_flag(FLAG_IS_TRIGGERED, lce);
+        return;
+    }
+
+    uint8_t is_trigger = 0;
+    const fixedQ48_t raw_grams = counts_to_grams(lce, sample);
+    const uint8_t is_overflow = overflows_int32(raw_grams);
+    if (!is_overflow) {
+        const fixedQ16_t filtered_grams = sosfilt(lce->sf
+            , (fixedQ16_t)raw_grams);
+        is_trigger = abs(filtered_grams) >= lce->trigger_grams_fixed;
+    } else {
+        // assume triggered in the case of an overflow
+        is_trigger = 1;
+        // While homing an overflow is an error
+        try_trigger_error(lce);
+    }
+
+    // latching trigger for QUERY_ENDSTOPS: dont clear until force is removed
+    uint8_t alredy_triggered = is_flag_set(FLAG_IS_TRIGGERED, lce);
+    if (alredy_triggered) {
+        const int64_t abs_int_grams = abs(round_fixedQ48(raw_grams));
+        uint8_t is_abs_trigger = abs_int_grams >= lce->trigger_grams;
+        is_trigger = is_abs_trigger || is_trigger;
+    }
+
+    // update trigger state
+    if (is_trigger) {
+        try_trigger(lce, ticks);
+    }
+    else {
+        // clear "live" trigger view for QUERY_ENDSTOPS
+        clear_flag(FLAG_IS_TRIGGERED, lce);
+    }
+}
+
+// Timer callback that monitors for timeouts
+static uint_fast8_t
+watchdog_event(struct timer *t)
+{
```

Comment:
I agree that we need to check for sensor failures.  However, I think that support should ultimately be a part of the trsync.c support (and not specific to loadcell).  We should also have this support in ldc1612 and likely other sensors in the future.  So, best to work out how we can do that generically.

I'll mockup some changes to trsync.c to demonstrate what I have in mind.

Reply:
That sounds good, I'm glad we agree. I'll leave this in and implement the associated fixes in the meantime.